compiler dependency: octave_statistics depends on cxx + new version 1…

cessenat opened the pull request:
….7.6 + new url

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->
New version compiles with recent compiler (gcc 13.3.0 from Ubuntu 24) and octave 9.4.0 at least.
This new version demands a new url location.

cessenat committed:
compiler dependency: octave_statistics depends on cxx + new version 1.7.6 + new url

cessenat committed:
Merge branch 'develop' into compiler_dep_octave_statistics

alalazo commented on repos/spack_repo/builtin/packages/octave_statistics/package.py:
```diff
@@ -12,9 +12,13 @@ class OctaveStatistics(OctavePackage, SourceforgePackage):
+    git = "https://github.com/gnu-octave/statistics/"
+    url = "https://github.com/gnu-octave/statistics/releases/download/release-1.7.6/statistics-1.7.6.tar.gz"
 
+    version("1.7.6", sha256="a518c50209e25e59742414c73955060f83c39c07b2d9a20b2fc8d13bd3106af3")
```
Did you check if the hash for v1.4.2 needs changes?

cessenat replied:
Hop, no, indeed 1.4.2 is not on github. Alike https://github.com/spack/spack-packages/pull/2244 it fails.
It seems that the packages that no longer update sourceforge files do not put old versions on github.
And I failed to make a `url_for_version` function that suits both.
It seems it is not so easy, for instance with otf2 package, check summing v2.3 one gets the checksum of v3.1:
```
root@3ab6b6d3802c:/# spack checksum otf2@2.3
==> Found 1 version of otf2
==> Fetching https://perftools.pages.jsc.fz-juelich.de/cicd/otf2/tags/otf2-3.1/otf2-3.1.tar.gz
    [100%]    5.58 MB @    5.9 MB/s

    version("3.1", sha256="09dff2eda692486b88ad5ee189bbc9d7ebc1f17c863108c44ccf9631badbada4")

```

cessenat replied:
It seems the def url_for_version fails to collaborate with spack checksum.
Here is another example:
```
root@3ab6b6d3802c:/# spack checksum parallel-netcdf@1.9.0
==> Selected 0 of 11 versions. Filtered by @1.9.0

==> Enter number of versions to take, or use a command:
    [c]hecksum  [e]dit  [f]ilter  [a]sk each  [n]ew only  [r]estart  [q]uit
action> c
root@3ab6b6d3802c:/# 
```
What do you think @alalazo ?

cessenat replied:
The combination of `sourceforge_mirror_path` and `def url_for_version` makes the `spack install `work fine with both legacy 1.4.2 release on sourceforge and new release 1.7.6 on github.
Still, spack checksum fails, but that is more general issue.
Could you review again please @alalazo ? Thanks.

cessenat committed:
spack install is fine with 1.7.6 but spack checksum is incorrect

cessenat committed:
Merge branch 'compiler_dep_octave_statistics' of ssh://github.com/cessenat/spack-packages into compiler_dep_octave_statistics
Update from remote

cessenat committed:
Merge branch 'develop' into compiler_dep_octave_statistics

cessenat committed:
typo corrected

cessenat committed:
Merge branch 'compiler_dep_octave_statistics' of ssh://github.com/cessenat/spack-packages into compiler_dep_octave_statistics
Update from remote

cessenat committed:
Merge branch 'develop' into compiler_dep_octave_statistics

cessenat committed:
Merge branch 'develop' into compiler_dep_octave_statistics

cessenat committed:
Merge branch 'develop' into compiler_dep_octave_statistics

cessenat committed:
Merge branch 'develop' into compiler_dep_octave_statistics

cessenat commented:
Hello @alalazo I need your help for that PR, I read "Requested changes must be addressed to merge this pull request." but  I have done the changes - or there is something I misunderstood. Now one can install the new release (1.7.6) as well as the former one (1.4.2) using a `url_for_version` method. Sorry for disturbing.

alalazo commented on repos/spack_repo/builtin/packages/octave_statistics/package.py:
```diff
@@ -13,8 +13,20 @@ class OctaveStatistics(OctavePackage, SourceforgePackage):
+            return f"https://sourceforge.net/projects/octave/files/Octave%20Forge%20Packages/Individual%20Package%20Releases/statistics--{version}.tar.gz"
+        else:
+            url = "https://github.com/gnu-octave/statistics/releases/download/release-{0}/statistics-{1}.tar.gz"
+            return url.format(version, version)
```
Since there is a single version downloaded from sourceforge, can't we just stick with a `github` url and add an explicit `url=` argument to v1.4.2 ? 

That would allow us to get rid of:
```suggestion
```
and make `spack versions` work too.

alalazo reviewed (changes requested):
I think we can simplify the recipe further:
```diff
diff --git a/repos/spack_repo/builtin/packages/octave_statistics/package.py b/repos/spack_repo/builtin/packages/octave_statistics/package.py
index 7867f0e4b5..66a32733fc 100644
--- a/repos/spack_repo/builtin/packages/octave_statistics/package.py
+++ b/repos/spack_repo/builtin/packages/octave_statistics/package.py
@@ -1,32 +1,26 @@
 # Copyright Spack Project Developers. See COPYRIGHT file for details.
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
-
 from spack_repo.builtin.build_systems.octave import OctavePackage
-from spack_repo.builtin.build_systems.sourceforge import SourceforgePackage
 
 from spack.package import *
 
 
-class OctaveStatistics(OctavePackage, SourceforgePackage):
+class OctaveStatistics(OctavePackage):
     """Additional statistics functions for Octave."""
 
     homepage = "https://octave.sourceforge.io/statistics/"
-    sourceforge_mirror_path = "octave/statistics-1.4.2.tar.gz"
     git = "https://github.com/gnu-octave/statistics/"
-    # url = "https://github.com/gnu-octave/statistics/releases/download/release-1.7.6/statistics-1.7.6.tar.gz"
+    url = "https://github.com/gnu-octave/statistics/releases/download/release-1.7.6/statistics-1.7.6.tar.gz"
 
     version("1.7.6", sha256="a518c50209e25e59742414c73955060f83c39c07b2d9a20b2fc8d13bd3106af3")
-    version("1.4.2", sha256="7976814f837508e70367548bfb0a6d30aa9e447d4e3a66914d069efb07876247")
+    version(
+        "1.4.2",
+        sha256="7976814f837508e70367548bfb0a6d30aa9e447d4e3a66914d069efb07876247",
+        url=f"https://sourceforge.net/projects/octave/files/Octave%20Forge%20Packages/Individual%20Package%20Releases/statistics--1.4.2.tar.gz"
+    )
 
     depends_on("octave-io")
     depends_on("cxx", type="build")
 
-    def url_for_version(self, version):
-        if version <= Version("1.4.3"):
-            return f"https://sourceforge.net/projects/octave/files/Octave%20Forge%20Packages/Individual%20Package%20Releases/statistics--{version}.tar.gz"
-        else:
-            url = "https://github.com/gnu-octave/statistics/releases/download/release-{0}/statistics-{1}.tar.gz"
-            return url.format(version, version)
-
     extends("octave@4.0.0:")

```

cessenat committed:
Update repos/spack_repo/builtin/packages/octave_statistics/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

cessenat committed:
Commiting @alalazo suggested changes

cessenat committed:
Merge branch 'develop' into compiler_dep_octave_statistics

cessenat committed:
Style demands a trailing comma

cessenat committed:
f-string is missing placeholders

cessenat commented:
Thanks @alalazo for your help ; I did the changes, just modified the "url=" for 1.4.2, added a comma and removed f"..." for style check. It now is fine with spack checksum, that is great. The "Requested changes" remains, still, if you validate I will apply the same technique for 2 other octave packages PR.
